packages/notifier/README.md

```diff
@@ -1,6 +1,263 @@
+updating a UI that doesn't care to hear about any older non-final values, as
+they are more stale. NotifierKit is appropriate even when this quantity changes
+quickly, as it will only communicate non-final values at the rate they're being
+consumed, bounded by the network round-trip time. All other non-final values
+are never communicated. The NotifierKit's lossy nature enables this
+optimization.
+
+The SubscriptionKit should be used for pub-sub operations, where every
+subscriber should see each published value starting with the starting point of
+their subscription. The producer can be described as
+the *publisher* and publishes iteration values with the `publication` using the
+`IterationObserver` API. The consumers can be described as *subscribers* and
+consume the published iteration values with the `subscription` using the
+`AsyncIterable` API. Since each published value will be sent to all
+subscribers, the SubscriptionKit should generally not be used with rapidly
+produced values.
+
+The values published using the publication define the original iteration. Each
+subscription has a starting point in that iteration and provides access to a
+suffix subset of that iteration starting at that starting point. The initial
+subscription created by the `makeSubscriptionKit()` call provides the entire
+iteration. Each subscription is a kind of `AsyncIterable` which produces any
+number of `AsyncIterators`, each of which advance independently starting with
+that subscription's starting point. These `AsyncIterators` are
+`SubsciptionIterators` which also have a `subscribe()` method. Calling a
+SubscriptionIterator's `subscribe()` method makes a `Subscription` whose
+starting point is that `SubscriptionIterator`'s current position at that time.
+
+Neither Alice nor Bob are good starting points to construct an example of
+`subscribe()` since their code uses only a `Subscription`, not
+a `SubscriptionIterator`. Carol's code is like Bob's except lower level, using
+a `SubscriptionIterator` directly. Where Bob uses `observeIteration` which
+takes an AsyncIterable`, Carol's uses the lower level `observeIterator` which
+takes an `AsyncIterator`.
+
+```js
+import { makePromiseKit } from '@agoric/promiseKit';
+
+const subscriptionIterator = subscription[Symbol.asyncIterator]();
+const { promise: afterA, resolve: afterAResolve } = makePromiseKit();
+
+const observer = harden({
+  updateState: val => {
+    if (val === 'a') {
+      afterAResolve(subscriptionIterator.subscribe());
+    }
+    console.log('non-final', val);
+  },
+  finish: completion => console.log('finished', completion),
+  fail: reason => console.log('failed', reason),
+});
+
+observeIterator(subscriptionIterator, observer);
+// eventually prints
+// non-final a
+// non-final b
+// finished done
+
+// afterA is an ERef<Subscription> so we use observeIteration on it.
+observeIteration(afterA, observer);
+// eventually prints
+// non-final b
+// finished done
+```
```

Comment:
I believe @michaelfig’s idea is captured by an lossy async iteration of lossless async iterations. An outer loop iterates over { state, delta } pairs, and the inner loop applies the deltas to the state.

I’d originally considered this idea around a connection-oriented abstraction, where the outer loop is driven by connection loss and reestablishment. But, another tool in this kit (not to be captured in this change) might buffer deltas and break out of the inner loop when the lossless buffer fills up. The buffer delta width might be configured to balance the trade-off of memory, network, and processor costs between using state deltas and full state syncs.

Reply:
> I believe @michaelfig’s idea is captured by an lossy async iteration of lossless async iterations. An outer loop iterates over { state, delta } pairs, and the inner loop applies the deltas to the state.

Great! That's exactly what I already did at https://github.com/Agoric/agoric-sdk/blob/async-iter-kit/packages/notifier/test/map-unum.js (though untested per disclaimer)

Reply:
> I’d originally considered this idea around a connection-oriented abstraction, where the outer loop is driven by connection loss and reestablishment. But, another tool in this kit (not to be captured in this change) might buffer deltas and break out of the inner loop when the lossless buffer fills up. The buffer delta width might be configured to balance the trade-off of memory, network, and processor costs between using state deltas and full state syncs.

An interesting thought. My experiment uses the heuristic of comparing the length of the accumulating deltas with the size  of a fresh snapshot.

> (not to be captured in this change)

Whew!